src/InMemoryMessageRepository.php

```diff
@@ -70,4 +72,16 @@ public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootV
 
         return $lastMessage instanceof Message ? $lastMessage->aggregateVersion() : 0;
     }
+
+    public function paginate(int $perPage, PaginationCursor | OffsetCursor $cursor): Generator
```

Comment:
I don't really understand the reason for the type expansion here, why would we want to type-hint against anything except the interface?

Reply:
The currently proposed design relies in capabilities of concrete implementations of the interface. This is often considered a design flaw, in this case I think it'd be OK. Effectively, the user of this tooling should use the correct type of cursor that matches the underlying storage mechanism. For the most part this will be the OffsetCursor, however, specialized cursors may choose to expose more methods for things like filtering, or non integer-type offsets (date ranges for example).

Reply:
Given that the vast majority of usage will be based on offset cursors, wouldn't it make sense to design the system that way? And if other implementations need, they can create their own replay functionality based on the (strong) example provided here?

Reply:
From a broad perspective, EventSauce has never tried to provide implementations for every possible situation, which is part of what keeps it lean and efficient. The current implementation of replay feels like it is trying to account for edge cases, which creates (IMO) a somewhat confusing and overly complex solution.

Reply:
In all honestly, this is essentially why I've held off supplying replaying for such a long time. The library has always provided interfaces that put the implementer in charge of these decisions, while focusing on orchestrations and interaction design. For replays, this is not really possible since resumable replays require "leaking" implementation details. The current design/abstraction does leak these details. The problem is that either the solution is ONLY usable for integer offsets OR we suffer from some design debt. The question is wether the costs of this design debt will hurt us, and in which situations it might. Right now, the current design allows offset based pagination and replays just fine. But perhaps you'd know of situations where this design choice may produce adverse affects?